### PR TITLE
When inserting/deleting lines, preserve RGB/256 attributes

### LINKED ISSUE #832: Vim background color renders incorrectly
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)
Microsoft Windows [Version 10.0.18362.86]

* What you're doing and what's happening: (Copy & paste specific commands and their output, or include screen shots)
In conhost or Windows Terminal, using both Vim.exe and WSL Vim, background color rendering becomes broken if a background color >16 or true colors is used.

With an example file, e.g. https://github.com/GilbertsHub/web-utils/blob/master/cgibinIntercept.sh:
```bash
vim -u NONE cgibinIntercept.sh +"hi Normal ctermbg=17" # for Windows: +"set t_Co=256"
# scroll down with CTRL-D
```

In general it is much easier to reproduce in WSL than Vim.exe, because it doesn't seem to occur after the initial scroll on Vim.exe and the effects are a bit different as described below. I can reproduce with any combination of conhost/terminal and Vim 8.1.883 on Ubuntu/8.0 on Debian/8.1 on Win32. Does not reproduce with neovim (v0.4.0-430-g8698830cb).

A few cases of example behavior:
   - Debian with conhost, Vim from official repository
```bash
$ sudo apt install vim
$ vim --version
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Sep 30 2017 18:21:38)
Included patches: 1-197, 322, 377-378, 550, 703, 706-707
```
vim-bg
Same behavior in the new terminal.

   - Vim.exe on conhost cmd.exe installed from https://github.com/vim/vim-win32-installer/releases/download/v8.1.1099/gvim_8.1.1099_x64.exe
vim-bg-conhost
Background color is wrong on the text part but correct otherwise (purple, although this seems to depend on active colorscheme) but is fixed after a Ctrl-L redraw.

   - Vim.exe on Terminal cmd.exe, same as above
vim-bg-terminal
Background color is again wrong, but the background highlight completely disappears after a Ctrl-L redraw.

* What's wrong / what should be happening instead:
Background rendered the whole width of the window, e.g. in wsltty:
vim-bg-wsltty

The effect is very similar to the issue https://github.com/microsoft/terminal/issues/375, however that one happens with bg color set to <16 (`darkBlue`). However I cannot reproduce that issue anymore, including the terminal bg change.

The issues for Win32 and Linux versions might be separate, since their effects are pretty different as described above.

I also haven't been able to reproduce this using VT sequences, but I hope the bug report is still useful.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Instead of doing our lossy hack to try and roundtrip the current attributes to legacy attrs and back, just straight up call ScrollRegion directly with the current attributes.

 
## References

 This may have some affect on #2553. It certainly doesn't fix it, but it might make that issue's solution easier.

## PR Checklist
* [x] Closes #832
* [x] I work here
* [x] Tests added/passed

## Detailed Description of the Pull Request / Additional comments

_Why didn't we do it this way in the first place?_ Simply put, the console has been a pretty heavily evolving codebase over the last four years. It wasn't until recently that `ScrollRegion` was updated to accept a full `TextAttribute` (as opposed to a `WORD`). Frankly, the ability to use a `TextAttribute` in that function means we can probably get rid of a bunch of similar hacks that do the same thing.  

## Validation Steps Performed

Ran the tests

Tried the simple repro from the issue.

Tried the more complicated vim repro from the OP.

### CODE REVIEW

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 1]
```diff
@@ -3451,6 +3453,73 @@ void ScreenBufferTests::ReverseLineFeedInMargins()
...
+    auto& cursor = si.GetTextBuffer().GetCursor();
+
+    TextAttribute expectedAttr{ si.GetAttributes() };
+    expectedAttr.SetBackground(gci.GetColorTableEntry(2));
+
+    // Set some scrolling margins
+    stateMachine.ProcessString(L"\x1b[1;3r");
+
+    // Set the BG color to the table index 2, as a 256-color sequence
+    stateMachine.ProcessString(L"\x1b[48;5;2m");
```
Comment (DHowett-MSFT) [inline src/host/ut_host/ScreenBufferTests.cpp]: actually, does 256-color `2` match 16-color `2`? I'd be concerned that this isn't testing something above the legacy attributes if so.

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 2]
```diff
@@ -167,6 +167,8 @@ class ScreenBufferTests
     TEST_METHOD(DeleteLinesInMargins);
     TEST_METHOD(ReverseLineFeedInMargins);
 
+    TEST_METHOD(DeleteLines256Colors);
```
Comment (miniksa) [inline src/host/ut_host/ScreenBufferTests.cpp]: What about InsertLines256Colors? You mentioned that in the PR title.

**Comment by j4james** [comment]
Would it not be worthwhile updating `DoSrvPrivateReverseLineFeed` at the same time, assuming it's just a matter of replacing the `ScrollConsoleScreenBufferWImpl` call with a direct call to `ScrollRegion`?

As for the testing, I've added a whole bunch of scrolling tests in PR #2505 which I think could easily be adapted to use 256 colors once this is merged. Although that PR is going to need to be rebased and updated first.

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 3]
```diff
@@ -167,6 +167,11 @@ class ScreenBufferTests
     TEST_METHOD(DeleteLinesInMargins);
     TEST_METHOD(ReverseLineFeedInMargins);
 
+    TEST_METHOD(InsertDeleteLines256Colors);
+    // TEST_METHOD(InsertDeleteLines256ColorsFrom16Table);
```
Comment (carlos-zamora) [inline src/host/ut_host/ScreenBufferTests.cpp]: Did you mean to uncomment these three `TEST_METHOD`s?

**Comment by kshji** [comment]
How this and same problem in WSL Ubuntu LTS 18.04 bash cmd shell was previously connected ?
Same problem start about same time also in bash cmd shell. After this Terminal update bash command shell still include this problem. 

But if you start Ubuntu bash shell using this Terminal, then vim works fine.

**Comment by zadjii-msft** [comment]
@kshji The Windows Terminal ships a standalone version of conhost.exe within its package that it uses to host console sessions. When you launch "ubuntu bash" directly, it's using the in-box conhost.exe, which was last updated whenever your OS was. When we make changes to the console code (like this PR), the Terminal is able to use those fixes before they're broadly available, because it takes a few weeks for those changes to be ingested by a Windows Insider build.

**Comment by kshji** [comment]
This process not work so nice as it has told. 

- I have used WLS Ubuntu 18.04 long time in my Windows
- suddenly I got the (and many others) the reverse problem, this unit not belongs to the Inside program
- I just got last update to my Windows
- still I have same problem
- too slow fixing process
- if this inside pipe is so clever, how does it was possible that so bad version conhost update has published in some "well and long tested" version ? But back to working version take so long ?
- also cut&paste newline problem has included this conhost update,

**Pull request merged**